examples/acoustic/fwi_operators.py

```diff
@@ -49,13 +66,36 @@ def __init__(self, model, src, damp, data, time_order=2, spc_order=6,
 class AdjointOperator(Operator):
     def __init__(self, model, damp, data, recin, time_order=2, spc_order=6, **kwargs):
         nrec, nt = data.shape
-        dt = model.get_critical_dt()
+        s, h = symbols('s h')
         v = TimeData(name="v", shape=model.get_shape_comp(), time_dim=nt,
-                     time_order=time_order, space_order=spc_order,
+                     time_order=2, space_order=spc_order,
                      save=False, dtype=damp.dtype)
         m = DenseData(name="m", shape=model.get_shape_comp(), dtype=damp.dtype)
         m.data[:] = model.padm()
         v.pad_time = False
+        # Derive stencil from symbolic equation
+        if time_order == 2:
+            Lap = v.laplace
+            Lap2 = 0
+            # PDE for information
+            # eqn = m * v.dt2 - Lap - damp * v.dt
+            dt = model.get_critical_dt()
+        else:
+            Lap = v.laplace
+            Lap2 = 1 / m * v.laplace2
+            # PDE for information
+            # eqn = m * v.dt2 - Lap - s**2 / 12 * Lap2 + damp * v.dt
+            dt = 1.73 * model.get_critical_dt()
+
+        # Create the stencil by hand instead of calling numpy solve for speed purposes
+        # Simple linear solve of a v(t+dt) + b u(t) + c v(t-dt) = L for v(t-dt)
+        stencil = 1.0 / (2.0 * m + s * damp) * \
+            (4.0 * m * v + (s * damp - 2.0 * m) *
+             v.forward + 2.0 * s ** 2 * (Lap + s ** 2 / 12.0 * Lap2))
+
+        # Add substitutions for spacing (temporal and spatial)
+        subs = {s: dt, h: model.get_spacing()}
```

Comment:
This is to avoid `solve` that is very slow for complex stencil. Im practice it only needs to solve 
```
a u(t+1) + b u(t) + c u(t-1) + L u(t) =0
```
but solve goes inside `L` for every term looking for simplification.